pyproject.toml

```diff
@@ -5,4 +5,12 @@
 # dynamically import GDAL via python's import system.  This behavior means
 # that we can provide a much easier build experience so long as GDAL is
 # available at runtime.
-requires = ['setuptools', 'wheel', 'setuptools_scm', 'numpy', 'cython']
+requires = [
+    'setuptools', 'wheel', 'setuptools_scm', 'cython',
+    # use minimum compatible numpy for each python version
+    # https://github.com/cython/cython/issues/4452
+    'numpy==1.16.5; python_version=="3.7"',
+    'numpy==1.17.3; python_version=="3.8"',
+    'numpy==1.19.3; python_version=="3.9"',
+    'numpy==1.21.2; python_version=="3.10"']
+build-backend = "setuptools.build_meta"
```

Comment:
Not quite sure how this is different from the [fallback option](https://pip.pypa.io/en/stable/reference/build-system/pyproject-toml/#fallback-behaviour), but it's included in [an example from the setuptools docs](https://setuptools.pypa.io/en/latest/build_meta.html#how-to-use-it).